### DISCUSSION TITLE
any reason `update_query` encode `Mapping` to `str` first then parse `str` to `Mapping`?

### DISCUSSION THREAD

**Discussion by trim21** [aio-libs/yarl#1012] [General]
I read the code and find that `update_query` would encode `Mapping`(or kwargs) to str and parse it to Mapping later, any reason don't just check mapping key and value instead of doing this?

**Comment by TayfurYldz** [comment]
On current yarl, the mapping path no longer does that string round-trip.

`update_query()` now handles a `Mapping` by starting from the already parsed query, updating a `MultiDict` directly, and serializing the result once:

```python
qm = MultiDict(self._parsed_query)
qm.update(in_query)
query = get_str_query_from_sequence_iterable(qm.items())
```

The string-input path is separate: only when the caller actually supplies a query string does yarl parse that string with `query_to_pairs()` before updating the existing query.

So the concern in the question was valid for the implementation you inspected, but the current implementation has the shape you were suggesting: mapping values remain structured through the merge and are encoded when the final query string is produced.

One reason serialization still has to happen at the end is that a URL stores its query in encoded URL form, while `URL.query` / the parsed-query cache exposes the structured `MultiDict` view. But there is no need for a mapping -> encoded string -> parsed mapping round-trip just to perform the update, and current yarl avoids it.
